core/src/main/scala/kafka/coordinator/TransactionLogManager.scala

```diff
@@ -17,15 +17,334 @@
+
+      PidKey(version, transactionalId)
+    } else if (version == CURRENT_TXN_KEY_SCHEMA_VERSION) {
+      // version 1 refers to txn status messages
+      val pid = key.get(TXN_KEY_SCHEMA_PID_FIELD).asInstanceOf[Long]
+
+      TxnKey(version, pid)
+    } else {
+      throw new IllegalStateException(s"Unknown version $version from the transaction log message")
+    }
+  }
+
+  /**
+    * Decodes the pid messages' payload and retrieves pid metadata from it
+    *
+    * @return a pid metadata object from the message
+    */
+  def readPidValue(buffer: ByteBuffer): PidMetadata = {
+    if (buffer == null) { // tombstone
+      null
+    } else {
+      val version = buffer.getShort
+      val valueSchema = schemaForPid(version)
+      val value = valueSchema.read(buffer)
+
+      if (version == CURRENT_PID_VALUE_SCHEMA_VERSION) {
+        val pid = value.get(PID_VALUE_SCHEMA_PID_FIELD).asInstanceOf[Long]
+        val epoch = value.get(PID_VALUE_SCHEMA_EPOCH_FIELD).asInstanceOf[Short]
+        val timeout = value.get(PID_VALUE_SCHEMA_TXN_TIMEOUT_FIELD).asInstanceOf[Int]
+
+        new PidMetadata(pid, epoch, timeout)
+      } else {
+        throw new IllegalStateException(s"Unknown version $version from the pid mapping message value")
+      }
+    }
+  }
+
+  /**
+    * Decodes the txn status messages' payload and retrieves transaction metadata from it
+    *
+    * @return a transaction metadata object from the message
+    */
+  def readTxnStatusValue(buffer: ByteBuffer): TransactionMetadata = {
+    if (buffer == null) { // tombstone
+      null
+    } else {
+      val version = buffer.getShort
+      val valueSchema = schemaForTxn(version)
+      val value = valueSchema.read(buffer)
+
+      if (version == CURRENT_TXN_VALUE_SCHEMA_VERSION) {
+        val stateByte = value.get(TXN_VALUE_STATUS_FIELD).asInstanceOf[Byte]
+        val state = TransactionMetadata.byteToState(stateByte)
+
+        val transactionMetadata = new TransactionMetadata(state)
+
+        val topicPartitionArray = value.getArray(TXN_VALUE_PARTITIONS_FIELD)
+
+        topicPartitionArray.foreach { memberMetadataObj =>
+          val memberMetadata = memberMetadataObj.asInstanceOf[Struct]
+          val topic = memberMetadata.get(PARTITION_SCHEMA_TOPIC_FIELD).asInstanceOf[String]
+          val partitionIdArray = memberMetadata.getArray(PARTITION_SCHEMA_PARTITIONS_FIELD)
+
+          val topicPartitions = partitionIdArray.map { partitionIdObj =>
+            val partitionId = partitionIdObj.asInstanceOf[Integer]
+            new TopicPartition(topic, partitionId)
+          }
+
+          transactionMetadata.addPartitions(topicPartitions.toSet)
+        }
+
+        transactionMetadata
+      } else {
+        throw new IllegalStateException(s"Unknown version $version from the transaction status message value")
+      }
+    }
+  }
+
+  // Formatter for use with tools to read transactional id to pid mapping history
+  class PidMessageFormatter extends MessageFormatter {
```

Comment:
I know we're just following the pattern used in `GroupMetadataManager`, but wouldn't it make more sense to handle both types in the same formatter? Is there any reason why they should be separate?